src/main/java/org/jabref/model/openoffice/style/CitedKey.java

```diff
@@ -0,0 +1,140 @@
+package org.jabref.model.openoffice.style;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Optional;
+
+import org.jabref.model.database.BibDatabase;
+import org.jabref.model.entry.BibEntry;
+import org.jabref.model.openoffice.ootext.OOText;
+
+/**
+ * Cited keys are collected from the citations in citation groups.
+ *
+ * They contain backreferences to the corresponding citations in {@code where}. This allows the
+ * extra information generated using CitedKeys to be distributed back to the in-text citations.
+ */
+public class CitedKey implements
+                      ComparableCitedKey,
+                      CitationMarkerNormEntry,
+                      CitationMarkerNumericBibEntry {
+
+    public final String citationKey;
+    private final List<CitationPath> where;
+
+    private Optional<CitationLookupResult> db;
+    private Optional<Integer> number; // For Numbered citation styles.
+    private Optional<String> uniqueLetter; // For AuthorYear citation styles.
+    private Optional<OOText> normCitMarker;  // For AuthorYear citation styles.
+
+    CitedKey(String citationKey, CitationPath path, Citation cit) {
+
+        this.citationKey = citationKey;
+        this.where = new ArrayList<>(); // remember order
+        this.where.add(path);
+
+        // synchronized with Citation
+        this.db = cit.getLookupResult();
+        this.number = cit.getNumber();
+        this.uniqueLetter = cit.getUniqueLetter();
+
+        // CitedKey only
+        this.normCitMarker = Optional.empty();
+    }
+
+    /*
+     * Implement ComparableCitedKey
+     */
+    @Override
+    public String getCitationKey() {
+        return citationKey;
+    }
+
+    @Override
+    public Optional<BibEntry> getBibEntry() {
+        return (db.isPresent()
```

Comment:
just Optional.of(db.get().entry)?

Reply:
No. `db` may be missing (no database, or not found in any), but we still want to
produce something, even if it is only "Unresolved(citationKey)". `db.get()` would throw in this case.

Would `db.map(e -> e.entry )` better?
Certainly shorter. Changed to `db.map(e -> e.entry)`